chapter_01/final_exercise_01.md

```diff
@@ -21,66 +21,129 @@
+
 25. **Q:** Can you define SerDe in the context of Hive?
+25. **A:** SerDe is the Serializer and the Deserializer which decides how hive stores and reads the data.
 
 ### Chapter 6: Apache ZooKeeper
 
 26. **Q:** What is Apache ZooKeeper and what role does it play in a distributed environment?
+26. **A:** Zookeeper is a distributed configuration manager  and coordinator that holds different configurations for different systems which allows different systems to coordinate with each other and allows for a single source of configuration change.
+
 27. **Q:** Can you explain the concept of znodes in ZooKeeper?
+27. **A:** ZNodes are Like data nodes which can have children, which in turn acts like files and directories.
+
 28. **Q:** How does ZooKeeper handle coordination and configuration management across distributed systems?
+28. **A:** Zookeeper handles coordination and configuration management using watches where clients can listen to changes, and atomicity and sessions which aid in the process of the coordination
+
 29. **Q:** What are ephemeral nodes and sequence nodes in ZooKeeper, and how are they used?
+29. **A:** *ephermal nodes* are nodes which last for as long as the session is kept alive which can be used for master / leader allocation and *sequence nodes* are nodes which have an assigned unique number, can be used for distributed counters.
+
 30. **Q:** How does ZooKeeper ensure data consistency and reliability across distributed nodes?
+30. **A:** Zookeeper ensures consistency and reliability by replication and only the primary node being the one interacted with, in adition to atomicity.
 
 ### Chapter 7: Apache HBase
 
 31. **Q:** What is Apache HBase?
 32. **Q:** What is a Row Key in HBase?
 33. **Q:** What is a Column Family in HBase?
-34. **Q:** How does HBase ensure data availability and fault tolerance?
+34. **Q:** How oes HBase ensure data availability and fault tolerance?
 35. **Q:** What is the role of ZooKeeper in an HBase environment?
 
 ### Chapter 8: Apache Kafka
 
 36. **Q:** What is Apache Kafka?
+36. **A:** Kafka is a distributed, replicated Streaming pub sub model which allows for event driven and event streaming architecture.
+
 37. **Q:** What are Topics in Kafka?
+37. **A:** Topics in kafka are a way to organize data, a topic is like a queue of data (could be multiple queues for distributed purposes and for performance improvements) which we can subscribe or publish to.
+
 38. **Q:** What are Partitions in Kafka?
+38. **A:** Partitions in kafka are a single queue which contain data inside of them, only one consumer group can be connected to the same partition
+
 39. **Q:** What is the role of a Broker in Kafka?
+39. **A:** The broker in kafka decides to which partition you will write or read and assigns partitions and stores data across multiple partitions (A kafka instance).
```

Comment:
Describe the flow if publishing and reading a message from kafka and where does the broker comes in play